custom_methods.rb

```diff
@@ -0,0 +1,145 @@
+# rubocop:disable Style/CaseEquality
+# rubocop:disable Metrics/ModuleLength
+module Enumerable
+  def change_self
+    return Array self if self.class == Range
+
+    self
+  end
+
+  def my_each
+    my_array = change_self
+    return to_enum unless block_given?
+
+    i = 0
+    while i < my_array.length
+      yield my_array[i]
+      i += 1
+    end
+    my_array
+  end
+
+  def my_each_with_index
+    my_array = change_self
+    return to_enum unless block_given?
+
+    i = 0
+    while i < my_array.length
+      yield my_array[i], i
+      i += 1
+    end
+    my_array
+  end
+
+  def my_select
+    my_array = change_self
+    return to_enum unless block_given?
+
+    arr = []
+    my_array.my_each { |x| arr.push(x) if yield x }
+    arr
+  end
+
+  def my_all?(word = nil)
+    my_array = change_self
+    if block_given?
+      my_array.my_each { |item| return false unless yield item }
+      true
+    elsif word
+      my_array.my_each { |item| return false unless word === item }
+      true
+    else
+      my_all? { |item| item }
+    end
+  end
+
+  def my_any?(regex = false)
+    my_array = change_self
+    if block_given?
+      my_array.my_each { |item| return true if yield item }
+      false
+    elsif regex
+      my_array.my_each { |item| return true if regex === item }
+      false
+    else
+      my_any? { |item| item }
+    end
+  end
+
+  def my_none?(regex = false)
+    my_array = change_self
+    if block_given?
+      my_array.my_each { |item| return false if yield item }
+      true
+    elsif regex
+      my_array.my_each { |item| return false if regex === item }
+      true
+    else
+      my_none? { |item| item }
+    end
+  end
+
+  def my_count(item = false)
+    my_array = change_self
+    counter = 0
+    if block_given?
+      my_array.my_each { |elem| counter += 1 if yield elem }
+      counter
+    elsif item
+      my_array.my_each { |elem| counter += 1 if elem == item }
+      counter
+    else
+      my_array.length
+    end
+  end
+
+  def my_map(proc = false)
+    my_array = change_self
+    arr = []
+    if block_given?
+      my_array.my_each { |item| arr.push(yield item) }
+      arr
+    elsif proc
+      my_array.my_each { |item| arr.push(proc.call(item)) }
+      arr
+    else
+      to_enum
+    end
+  end
```

Comment:
#### my_map

[OPTIONAL]

- [ ] In the specifications of the project in the [Odin project](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks) # 12, you were asked to modify your my_map method to take either a proc or a block. And if both a proc and a block are given, only execute the proc. The real map doesn't have this feature, but you were asked to have it so I strongly recommend working on it

Given 
`my_proc = Proc.new {|num| num > 10 }`

Result
> puts [18, 22, 5, 6] .my_map(my_proc) {|num| num < 10 }  # true true false false